### ISSUE TITLE
Way to specify flycheck checker and integration with virtual environments in python

### ISSUE THREAD

**Issue by vikigenius** [hlissner/doom-emacs#358]
Is there any plan to provide integration with python virtual environments like conda environments or virtualenvs?

For now, is there a way to set the flycheck checker executable to avoid errors arising from virtual environments?

When i try to set the executable manually, i am getting void function errors for `pkg-info-version-info"` declared in flycheck.

**Comment by hlissner** [comment]
The `pkg-info-version-info` should be fixed as of https://github.com/hlissner/doom-emacs/commit/b741d77e7897707a8739f2eae6b0ba083c5a0fad.

However, I'm not entirely sure what integration for virtualenv/conda would entail, since I don't use them (I've always used pyenv). I'm interested in learning more about conda, so I'd appreciate any insight you can give me.

**Comment by vikigenius** [comment]
I am working on it. I use conda mainly because it is the best there is for scientific packages and can install packages without any hassle and doesn't require compiling at all, just binaries. I work a lot in ML and Data Analysis and it's a god send for me. The virtual environments are just an added bonus.

There is already an existing package [conda.el](https://github.com/necaris/conda.el) which does provide an integration with anaconda-mode, but i am not really happy with it and i don't like how they seem to be activating every time a buffer switch happens, I am working on implementing something on my own, but i am new to emacs/lisp so i am somewhat struggling and it may take time.

I have a question about the autoloading issue here though? What determines which autolads are auto loaded while lazy loading? Can you eli5, why this pkg-info-version-info is not being loaded when needed by flycheck?

**Comment by vikigenius** [comment]
Expanding on conda, it is basically a package manager, that is included with the anaconda distribution. It's virtual environment basically combines the functionality of pip and virtualenvs. 

Conda environments provide a great way to completely isolate your project environment from your system, and conda is crucial here because it can also install system dependencies when you install a package with conda, so everything required for your python project including libraries and the python executable themselves are in one place and the whole config can be reliably exported to a file and can be replicated on another machine.

I think the main issue of integration is getting packages like flycheck, anaconda-mode etc. to use your conda  (or any such virtual environment's like virtualenv) environment's python and packages instead of the default system ones. So one main step that a lot of these integration packages like [conda.el](https://github.com/necaris/conda.el) for conda and [pyvenv](https://github.com/jorgenschaefer/pyvenv) for virtualenvs do is prepend the environment location to the exec path variable in the shell.

Most of the other implementation details is probably concerned about detecting the right environment to use for a project by looking at some special file in project root and providing key bindings to switch environments and stuff like that

**Comment by fuxialexander** [comment]
@vikigenius I just set up `conda.el` and `lsp-python` yesterday, and so far I like it. In order to avoid the hook of switch-buffer, I think you can just turn off `conda-env-autoactivate-mode`, which is just a minor mode that set hook to `switch-buffer`. You can also rewrite the minor mode to hook it not to switch-buffer, but to other function.

**Comment by vikigenius** [comment]
@fuxialexander , what function do you suggest, hooking it with. I like the functionality of auto activating environments based on the environment.yml files.

And how good is lsp-python? I have heard about the language server protocol project. And it sounds similar to what google is doing with kythe, but kythe itself is not really in a good stage right now.
So how good is Microsoft's LSP and how does it compare to existing autocomplete solutions?

**Comment by hlissner** [comment]
> What determines which autolads are auto loaded while lazy loading? Can you eli5, why this pkg-info-version-info is not being loaded when needed by flycheck?

@vikigenius Doom avoids calling `package-initialize` when byte-compiled (mostly for performance reasons), which would otherwise have loaded package autoload files, like flycheck's, where an autoload declaration for `pkg-info-version-info` exists. 

Since I prefer declaring autoloads myself and this decision only truly affects a few plugins (haskell-mode, auctex and flycheck), I accept that compromise.

**Comment by hlissner** [comment]
This should be fixed as of the latest `develop`, so I'll close this. Please refer to #736 for a follow up on virtualenv tooling (pyenv, pyvenv, pipenv & conda).
